### DISCUSSION TITLE
Idea: Edit photo in local Photoshop application

### DISCUSSION THREAD

**Discussion by djthorpe** [photoprism/photoprism#2551] [Ideas]
I would be keen to have the ability to edit photos that are stored in photoprism in a local desktop application like Photoshop, and I was wondering if there are other people who are keen on this idea and/or already implementing it. A possible solution would be:

1. Have a "connector" application running on your local desktop machine, which links to your Photoprism instance. This would run in the background (ultimately it would have no user interface other than configuration) and connect to photoprism through websockets - announce "here I am, and this is my unique session identifier"
2. In photoprism, you would select the photos you want to "edit". It's not totally clear to me how you would choose an original vs an edited version of your photo, but suggestions would be great. You'd then click a button to send the photos to your connector through the opened websocket with the session identifier, webdav location URL and list of photos to edit.
3. The connector would download the photos from photoprism through WebDav to a temporary desktop location, and fire up Photoshop. Then it would somehow monitor the editing process (maybe wait until Photoshop quits?)
4. For any updated photos, the connector would re-upload your photos via WebDav to your originals folder, ask photoprism to index them, and somehow link it to the "stack" of your photos, making the edited photo the current one on the top of the stack.
5. The connector application would then delete any temporary images on your local desktop.

WebSockets connection would need to be maintained and re-established if the connection is broken.

Has someone proposed this sort of thing before? I would be open to suggestions on how to do this, since I'm not totally across Photoprism and the API at the moment. However, I can write a connector as a proof of concept.

Let me know what you think.

**Comment by graciousgrey** [comment]
A few thoughts on this:

I like the idea of being able to select photos in PhotoPrism to open them in external tools.
As you mention, one problem is that it is currently only possible to select photos, not files. This could be solved by default settings like "always open raw" or "open all associated files" or by a selection dialog in the web UI. 

I don't think it is necessary to upload or download photos to a temporary location. Unless PhotoPrism is running in public mode, the photos in the Originals folder can be accessed from your local desktop using webdav (https://docs.photoprism.app/user-guide/sync/webdav/). This would allow you to open the raw file directly from the originals folder in Photoshop, and have Photoshop save the new JPG version directly to the originals folder with a naming scheme to stack the file the next time it is indexed.

Somehow related:
- https://github.com/photoprism/photoprism/issues/1518
- https://github.com/photoprism/photoprism/issues/1028

**Comment by djthorpe** [comment]
OK thank you Theresa for your thoughts. I'll write a demo version to gain some user feedback (will take me a few weeks I think) and determine feasability. As you mention you could use such an mechanism to also export photos to your local desktop too, which would require a bit more functionality in photoprism-connect. I'd be keen to understand how stacked photos are implemented eventually to see how different "versions" of edited photos can be grouped together.

For this idea to work, the components would be:

### photoprism-connect

A small command-line application which persists on your local host (which has Photoshop installed). I think the configuration will need:

1. URL of photoprism instance API
2. Your username/password or other authentication details needed to communicate
3. An array of image editing applications that can be used to edit the images. For Linux-based machines just a map of unique application names to a set of paths to applications. For Macintosh-based it could be a set of application bundle ID's instead.
4. A unique session ID is generated when the application is loaded, and the hostname or other "friendly name" of your computer.

### photoprism-instance

This would eventually be your photoprism instance, but for demonstration/user feedback purposes it would be a separate server-based application. There would be three API calls:

1. `CONNECT /api/connect` Register Connect Application - websocket connection maintained. Request parameters are session ID and the array of application names.
2. `GET /api/connect` Return a set of currently opened desktop connections, with the applications for each connection
3. `POST /api/connect` Perform Edit operation - the request parameters would be the unique photo ID's to edit, the session ID of the computer to edit them on, and potentially the name of the application to use for editing (could default to the main application ie, Photoshop, otherwise).

Ultimately API calls 2 and 3 would communicate with your UI when either clicking on a single photo and choosing to "edit in photoshop" or on multiple photos. I'd also need some sort of API call (which may already exist) to re-index an existing photo by path or unique photo id.

The information transmitted from photoprism-instance to  photoprism-connect would be:

* An array of the photos to edit. At minimum, this would be the path of the photo and potentially the unique photo id.
* The URL of the webdav connection to locate the photos
* The unique name of the application to open

### Chrome Extension

I guess a chrome extension could be written to simulate how this would all work together seemlessly, adding a menu item when clicking "edit" on a photo.
